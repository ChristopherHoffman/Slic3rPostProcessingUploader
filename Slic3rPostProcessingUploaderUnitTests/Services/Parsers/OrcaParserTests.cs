using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Slic3rPostProcessingUploader.Services.Parsers;
using Snapshooter.MSTest;


namespace Slic3rPostProcessingUploaderUnitTests.Services.Parsers
{
    [TestClass]
    public class OrcaParserTests
    {
        [TestMethod]
        public void ShouldReturnAnEmptySettingWhenGcodeIsEmpty()
        {
            var parser = new OrcaParser("");
            var result = parser.ParseGcode("");
            Assert.IsNotNull(result);
            Assert.IsNotNull(result.settings);
            Assert.AreEqual(0, result.settings.estimated_print_time_seconds);
            Assert.AreEqual(0, result.settings.material_used_mg);
            Assert.AreEqual("Settings:", result.settings.note);
            Assert.AreEqual("OrcaSlicer", result.Slicer);
            Assert.AreEqual("Unknown", result.CuraVersion);
        }

        [TestMethod]
        public void ShouldReturnExpectedValuesWhenGivenFullGcode()
        {
            var parser = new OrcaParser("");
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Snapshot.Match(result);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithNoReplacements()
        {
            string template = "Settings:";

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("Settings:", result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithASingleReplacement()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                """;

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithMultipleReplacements()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                    First Layer Height: {{first_layer_height}}
                    Wall Loops: {{wall_loops}}
                    Top Shell Layers: {{top_shell_layers}}
                    Bottom Shell Layers: {{bottom_shell_layers}}
                    Sparse Infill Density: {{sparse_infill_density}}
                """;

            var parser = new OrcaParser(template);
            var result = parser.ParseGcode(OrcaParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                    First Layer Height: 0.200
                    Wall Loops: 3
                    Top Shell Layers: 4
                    Bottom Shell Layers: 3
                    Sparse Infill Density: 15%
                """, result.settings.note);
        }

    }
}
