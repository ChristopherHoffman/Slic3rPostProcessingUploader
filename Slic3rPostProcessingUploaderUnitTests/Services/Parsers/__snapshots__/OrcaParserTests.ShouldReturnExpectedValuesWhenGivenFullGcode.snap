{
  "Slicer": "OrcaSlicer",
  "CuraVersion": "2.2.0-rc",
  "PluginVersion": null,
  "settings": {
    "note": "Settings:\n\n  first_layer_height: 0.200\n  layer_height: 0.2\n  wall_loops: 3\n  top_shell_layers: 4\n  bottom_shell_layers: 3\n  sparse_infill_density: 15%\n  travel_speed: 150\n  nozzle_diameter: 0.4\n  filament_type: PLA\n  filament_diameter: 1.75\n  nozzle_temperature: 205\n  first_layer_temperature: 205\n  first_layer_bed_temperature: 60\n  brim_width: 5\n  raft_layers: 0\n  pressure_advance: 0.02\n  print_settings_id: 0.20mm Standard @MyMarlin\n  printer_settings_id: JGMaker R1",
    "print_name": null,
    "time_stamp": null,
    "estimated_print_time_seconds": 1319,
    "material_used_mg": 4530,
    "Snapshot": "iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAACoUlEQVR4Ae1YTU8TURR9CzcajQGhdN\npCQRITFyYudOHGmLgwJEYXhhjQICVChUiVqkFBoIHwtQBCWLHip/BT+lMec4fcZijvzdz3MfNmEk5y\n06Yzac85957b1zJ2Azl63r5pFaYnzlgeUd353x49PeRQ8Pz2o9FXLA8A15F4d8E1lmWUfi2cy8hjwT\n23HvQOsywhynVRwUjde/F8mmUBquS7u8FcIhxU3XIScIrrD0/2lISkFvA410eOd3n530/e93mce426\ncjcSCzjFdSBfXJjhfZMfeKk5HwhQ7UQi3aCQD1z3iRdmJgMhOEbV3TVlAdYCDuGijgwQh8fu68MHW1\noCjANOXY8wKuB8dX9DeN1EgPZIVVaW2qYfGicAOje0vUp+n8H1P22ygOK3Gh9sLVsRIMoAkIeRK87X\nSCHHfCkJgCr//m5VQGfF+mREeREJxfu1BGCZzjG4jK7Lwn5NuJ8puBfJGwmAqqw0tQV4P+odErhiqa\n5bE2DSDW9xljwy3a5bF2Az4BTXExGAAdc5JogKjx+pCsDSPSpAUVxPXIDOuo0amUJtgg/Uv6QrQCXg\nUUEF8tW9jfQ7QAk4JaheYy4opwKwwgGPW4/o/sjRtpsMRHWDGtQ4950IgKKQB/eHNv+620KmAvqnPv\nKSv8lyK8D594CJgOLiVz4wN9U5pUZlJnMCYPYrq0uXz0N/BuDPVG0B/Z/G22kI8BqzvoDmlddk5Hve\nj9F/UgLuPnvaSlLApfvN4DFuzO48edxiukhKgMh9Y9dl6H03dm5bAJCXHdqg7r9+af+fa1sC4PxfXm\n4k67oMqt2QCXBCHqES8LiAInGjoOrChoDUXJchbqQSW4+2oSLAuesyiLqRG/JhiATkgngYeJ7KJXkE\nrNtMBTWLuABl35zINAlR9wAAAABJRU5ErkJggg==",
    "file_name": null
  }
}
